drivers: sensor: Add tmd2620

pefech opened the pull request:
Original PR: #50978 was closed because it was too long on stale. This PR includes all requested fixes

This PR adds the driver for the IR proximity sensor tmd2620 made by ams-osram including a
generic proximity sensor polling sample with a device tree overlay for the nrf52840dk board that has the tmd2620 hooked up.
I tested the driver as well as the sample.

You can find datasheets of the sensor here: [TMD2620 datasheet](https://ams.com/documents/20143/36005/TMD2620_DS000461_3-00.pdf)

pefech committed:
drivers: sensor: Add tmd2620 driver

Adds tmd2620 driver and devicetree bindings to work in
trigger and polling mode supporting Power management.

Signed-off-by: Peter Fecher <p.fecher@phytec.de>

pefech committed:
samples: sensor: Add proximity sensor sample

Adds generic sample for proximity sensors.

Signed-off-by: Peter Fecher <p.fecher@phytec.de>

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,425 @@
+
+	LOG_DBG("Configuring Interrupt.");
+
+	if (!device_is_ready(config->int_gpio.port)) {
```
```suggestion
	if (!gpio_is_ready_dt(&config->int_gpio)) {
```

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,425 @@
+	uint8_t tmp;
+#endif
+
+	if (!device_is_ready(config->i2c.bus)) {
```
```suggestion
	if (!i2c_is_ready_dt(&config->i2c)) {
```

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,425 @@
+
+#define TMD2620_INIT_INST(n) \
+	struct tmd2620_data tmd2620_data_##n; \
+	static struct tmd2620_config tmd2620_config_##n = { \
```
```suggestion
	static const struct tmd2620_config tmd2620_config_##n = { \
```

bbilas commented on drivers/sensor/tmd2620/tmd2620.h:
```diff
@@ -0,0 +1,176 @@
+	const struct device *dev;
+	uint8_t pdata;
+
+	#ifdef CONFIG_TMD2620_TRIGGER
```
```suggestion
#ifdef CONFIG_TMD2620_TRIGGER
```

bbilas commented on drivers/sensor/tmd2620/tmd2620.h:
```diff
@@ -0,0 +1,176 @@
+void tmd2620_work_cb(struct k_work *work);
+
+int tmd2620_attr_set(const struct device *dev,
+						enum sensor_channel chan,
```
The indentation looks strange.

pefech commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -646,3 +646,17 @@ test_i2c_as5600: as5600@54 {
 	compatible = "ams,as5600";
 	reg = <0x54>;
 };
+
+test_i2c_tmd2620: tmd2620@65 {
+	compatible = "ams,tmd2620";
+	reg = <0x65>;
```
Here I used 0x65 as sequential address. This would be the right address if the entry before had the correct address.

pefech replied:
Opened a PR to correct the address of the entry before the tmd2620: #54264

pefech commented:
I applied all change requests

pefech commented:
@MaureenHelm is there any chance you could re-review this when you get a chance?

MaureenHelm reviewed (changes requested):
A few nits, but otherwise looks good

pefech commented on drivers/sensor/tmd2620/tmd2620_trigger.c:
```diff
@@ -0,0 +1,97 @@
+	const struct device *dev = data->dev;
+
+	if (data->p_th_handler != NULL) {
+		data->p_th_handler(dev, &data->p_th_trigger);
```
You are completely right. I've changed it, thanks

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,413 @@
+
+	if (!gpio_is_ready_dt(&config->int_gpio)) {
+		LOG_ERR("Interrupt GPIO device not ready");
+		return -EIO;
```
```suggestion
		return -ENODEV;
```

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,413 @@
+
+	if (gpio_add_callback(config->int_gpio.port, &data->gpio_cb) < 0) {
+		LOG_ERR("Failed to set GPIO callback");
+		return -EIO;
```
Don't use `-EIO` everywhere, please propagate the rc from functions instead.

pefech replied:
Do I understand you correctly, I should return the return value of the error-producing function?
In this example the return value of `gpio_add_callback(config->int_gpio.port, &data->gpio_cb)` ?

bbilas replied:
<pre>
ret = gpio_add_callback(config->int_gpio.port, &data->gpio_cb);
if (ret < 0)  {
	LOG_ERR("Failed to set GPIO callback");
	return ret;
}
</pre>

This way we would know the actual return code, instead of the hardcoded `EIO`;

pefech replied:
Okay, I understand. I'll change that

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,413 @@
+
+#ifdef CONFIG_TMD2620_TRIGGER
+	data->work.handler = tmd2620_work_cb;
+
```
Useless nl.

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,413 @@
+
+static int tmd2620_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+
```
ditto

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,413 @@
+		return -EINVAL;
+	}
+
+	tmp = tmp | config->proximity_led_drive_strength;
```
```suggestion
	tmp |= config->proximity_led_drive_strength;
```

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,413 @@
+
+	if (!i2c_is_ready_dt(&config->i2c)) {
+		LOG_ERR("I2C bus not ready!");
+		return -EIO;
```
```suggestion
		return -ENODEV;
```

bbilas commented on drivers/sensor/tmd2620/tmd2620_trigger.c:
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2022 PHYTEC Messtechnik GmbH
```
```suggestion
 * Copyright (c) 2023 PHYTEC Messtechnik GmbH
```

bbilas commented on dts/bindings/sensor/ams,tmd2620.yaml:
```diff
@@ -0,0 +1,81 @@
+  int-gpios:
+    type: phandle-array
+    required: true
+    description: Interrupt pin.
```
```suggestion
    description: |
       The interrupt pin of TMD2620 is open-drain, active low.
```

bbilas commented on drivers/sensor/tmd2620/Kconfig:
```diff
@@ -0,0 +1,33 @@
+# Copyright (c) 2022 PHYTEC Messtechnik GmbH
```
```suggestion
# Copyright (c) 2023 PHYTEC Messtechnik GmbH
```

pefech commented:
@bbilas I've removed all unnecessary new lines, I found, renewed the head line of every file, removed the not used property and I'm returning now the original return code instead of -EIO where useful

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,423 @@
+	tmp = TMD2620_ENABLE_PEN | TMD2620_ENABLE_PON;
+	ret = i2c_reg_update_byte_dt(&config->i2c, TMD2620_ENABLE_REG, tmp, tmp);
+	if (ret) {
+
```
ditto

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,423 @@
+
+	const struct tmd2620_config *config = dev->config;
+	uint8_t en_bits = TMD2620_ENABLE_PON;
+	int ret = 0;
```
```suggestion
	int ret;
```

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,423 @@
+	case PM_DEVICE_ACTION_RESUME:
+		ret = i2c_reg_update_byte_dt(&config->i2c, TMD2620_ENABLE_REG, en_bits, en_bits);
+		if (ret) {
+
```
Useless nl.

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,423 @@
+	case PM_DEVICE_ACTION_SUSPEND:
+		ret = i2c_reg_update_byte_dt(&config->i2c, TMD2620_ENABLE_REG, en_bits, 0);
+		if (ret) {
+
```
ditto

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,423 @@
+			return ret;
+		}
+		break;
+
```
ditto

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,423 @@
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		ret = i2c_reg_update_byte_dt(&config->i2c, TMD2620_ENABLE_REG, en_bits, en_bits);
```
Can't you just use `TMD2620_ENABLE_PON` instead of `en_bits` variable?

bbilas commented on drivers/sensor/tmd2620/tmd2620.h:
```diff
@@ -0,0 +1,178 @@
+
+static void tmd2620_setup_int(const struct tmd2620_config *config, bool enable)
+{
+
```
Useless nl.

bbilas commented on drivers/sensor/tmd2620/tmd2620_trigger.c:
```diff
@@ -0,0 +1,95 @@
+	LOG_DBG("Setting sensor attributes.");
+
+	const struct tmd2620_config *config = dev->config;
+	int ret = 0;
```
```suggestion
	int ret;
```

bbilas commented on drivers/sensor/tmd2620/tmd2620_trigger.c:
```diff
@@ -0,0 +1,95 @@
+
+	const struct tmd2620_config *config = dev->config;
+	struct tmd2620_data *data = dev->data;
+	int ret = 0;
```
```suggestion
	int ret;
```

bbilas commented on drivers/sensor/tmd2620/tmd2620_trigger.c:
```diff
@@ -0,0 +1,95 @@
+
+	tmd2620_setup_int(dev->config, false);
+
+	if (trigg->type == SENSOR_TRIG_THRESHOLD) {
```
It is super nested condition, can we make it a bit nicer? For example:

<pre>
if (trigg->type != SENSOR_TRIG_THRESHOLD) {
    return -ENOTSUP;
}

if (trigg->chan != SENSOR_CHAN_PROX) {
   return -ENOTSUP;
}

data->p_th_trigger = trigg;
data->p_th_handler = handler;

ret = i2c_reg_update_byte_dt(&config->i2c, TMD2620_INTENAB_REG,
						TMD2620_INTENAB_PIEN, TMD2620_INTENAB_PIEN);
if (ret < 0) {
	return ret;
}
</pre>

bbilas commented on dts/bindings/sensor/ams,tmd2620.yaml:
```diff
@@ -0,0 +1,80 @@
+# Copyright (c) 2022 Phytec Messtechnik GmbH
```
```suggestion
# Copyright (c) 2023 Phytec Messtechnik GmbH
```

bbilas commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,423 @@
+		return -ENODEV;
+	}
+
+	gpio_pin_configure_dt(&config->int_gpio, GPIO_INPUT | config->int_gpio.dt_flags);
```
btw. `config->int_gpio.dt_flags` is superflous there.

pefech commented:
@bbilas, I've changed everything you mentioned

Thank you all for your reviews and your patience :)
